### DISCUSSION TITLE
IPv6 connections to peers

### DISCUSSION THREAD

**Discussion by cross** [Chia-Network/chia-blockchain#8552] [Support]
chia-blockchain 1.2.3, Ubuntu 20.  installed from the .deb distribution here.

I have a harvester I'm trying to connect to a farmer peer in the same building.  I have other harvesters connected via IPv4.  This newest harvester is on a segmented IPv4 network, so I'm specifying the IPv6 address of the farmer as follows in my config.yml:

```
harvester:
  chia_ssl_ca:
    crt: config/ssl/ca/chia_ca.crt
    key: config/ssl/ca/chia_ca.key
  farmer_peer:
    host: 2000:3000:4000:5000::101
    port: 8447
  logging: *id001
  network_overrides: *id002
```

But, when I start `chia start harvester -r` the following appears on my terminal:
 
```
chia_harvester: started
user@host:~$ Traceback (most recent call last):
  File "chia/server/start_harvester.py", line 57, in <module>
  File "chia/server/start_harvester.py", line 53, in main
  File "chia/server/start_service.py", line 255, in run_service
  File "asyncio/runners.py", line 44, in run
  File "asyncio/base_events.py", line 616, in run_until_complete
  File "chia/server/start_service.py", line 249, in async_run_service
  File "chia/server/start_service.py", line 171, in run
  File "chia/server/start_service.py", line 145, in start
  File "chia/server/start_service.py", line 146, in <listcomp>
  File "chia/server/reconnect_task.py", line 12, in start_reconnect_task
socket.gaierror: [Errno -9] Address family for hostname not supported
[1752] Failed to execute script start_harvester
```

Do I have the syntax wrong, or does IPv6 connections to farmer have an issue?

(`netstat -an | grep 8447` on the farmer shows `tcp6       0      0 :::8447                 :::*                    LISTEN` amongst `tcp LISTEN` and ESTABLISHED connections via IPv4.  And I've verified I can establish the tcp connection using `telnet` to that IPv6 address, port 8447.)

**Comment by emlowe** [comment]
From looking at the code, I can confirm this won't work, as the code is using python gethostbyname call here (in start_reconnect_task) which only supports IPv4. I will create an issue for this

**Comment by cross** [comment]
Okay.  Now that PR #8765 is in, the problem I described above is fixed, so that I can connect to a farmer via IPv6 address.  But, names that resolve as IPv6 only will still fail.  Additionally I'd like to be able to prefer IPv6 for hostnames that resolve to both.

I propose that the config.yml have a boolean for each of "enable IPv6 address use/resolution", and "prefer IPv6 addresses".  Then it would be easy to modify the code in the places I know of that call `gethostbyname` to do the right thing.  I've already prototyped a method to do this based on those two booleans.

I can contribute this as a PR, but I need guidance for two things:

1. Are new config variables acceptable?  If so, let me know context and names, and point me to where that lives in the code.
2. In the case that IPv6 is enabled, but prefer IPv6 is False, should that imply "prefer IPv4", or should it just use whatever the resolver returns first?

**Reply by jack60612** [reply]
> Okay. Now that PR #8765 is in, the problem I described above is fixed, so that I can connect to a farmer via IPv6 address. But, names that resolve as IPv6 only will still fail. Additionally I'd like to be able to prefer IPv6 for hostnames that resolve to both.
> 
> I propose that the config.yml have a boolean for each of "enable IPv6 address use/resolution", and "prefer IPv6 addresses". Then it would be easy to modify the code in the places I know of that call `gethostbyname` to do the right thing. I've already prototyped a method to do this based on those two booleans.
> 
> I can contribute this as a PR, but I need guidance for two things:
> 
> 1. Are new config variables acceptable?  If so, let me know context and names, and point me to where that lives in the code.
> 2. In the case that IPv6 is enabled, but prefer IPv6 is False, should that imply "prefer IPv4", or should it just use whatever the resolver returns first?

If you want just contribute it as a draft pr and we can provide easy feedback that way

**Reply by cross** [reply]
Well, I need to know the questions above before I know how to code it.  The # 2 above can be changed easily after review, but # 1 is pretty structural.  If you're accepting of the idea of the two new config variables, I would like a pointer to where that is defined in the code, and at what level they should exist.  There are very few top-level config vars, but that seems likely the best place for these IMO.

**Reply by jack60612** [reply]
> Well, I need to know the questions above before I know how to code it. The # 2 above can be changed easily after review, but # 1 is pretty structural. If you're accepting of the idea of the two new config variables, I would like a pointer to where that is defined in the code, and at what level they should exist. There are very few top-level config vars, but that seems likely the best place for these IMO.

If you want just set them as static variables for now.
I'll. find it in a bit

**Comment by cross** [comment]
Tagging @jack60612  and @hoffmang9  based on their input to #8765

**Reply by jack60612** [reply]
Let me get back to you in the morning, i'm heading off for the night.

**Comment by emlowe** [comment]
I don't think you need two config options. I think a single option just for "prefer IPv6 addresses" is better - the code should always look up both IPv4 and IPv6, and then finding IPv6 use that first if that pref is turned on, otherwise use IPv4. If the names only resolves to IPv4, use that regardless, if the name only resolves to IPv6, ditto.

new config options are fine - you can add them into initial-config.yaml.

**Reply by cross** [reply]
That's fine, thanks @emlowe.  Some people like to be able to disable IPv6 entirely in applications, but not I, so I'm happy to not do that.  Hopefully "prefer IPv6: false" will be adequate for anyone with that desire.
I'll lean this way when I get to this, which sadly I haven't found time for yet.  Hopefully within the next week.

**Reply by cross** [reply]
Hey Earle.  Adding it to the `initial-config.yaml` is clear, but is there code that already handles putting that into configs that don't have it, or is there a way that a default value is set in memory somehow if it's not in the config?

**Comment by cross** [comment]
I wrote a resolver function in `chia/util/network.py`, and replaced the earlier changes around `gethostbyname` to use that.  A problem I'm having however after trying to add the config variable is, can I from within that standalone utility method get ahold of the config?  Or, the root_path so that I can use `load_config` ?
@emlowe @jack60612

**Reply by cross** [reply]
Okay.  I found DEFAULT_ROOT_PATH, which I can use inside of my resolver method (`get_host_addr`) in `chia/util/network.py`.  The only outstanding question I have is whether I can mock the config in my unit-test.  For now, I have an optional arg to my method that can force a preference, but the optional arg to the method just for testing is ugly.
I'll be putting up a PR shortly where we can discuss.  It's at least mostly there as a first pass now.

**Comment by cross** [comment]
PR #8861

**Reply by hoffmang9** [reply]
The pattern we've been using since mainnet launch is to check for existence of the new configuration element. If it does not exist then use the default. Depending on the item you would also go ahead and save the new default. This is one of those things that we want to save the default of use IPv6 to configurations if it is not specified the other way.

**Reply by cross** [reply]
Sounds good.  Can you point me to an example of that, where a new config element was introduced and saved?  Just want to make sure the way I store the default in the code and write it all out is succinct and easy to identify.  (I assume I'll set/ref the default in my new method, since that's the only place I'm loading it out of config at the moment.)
I definitely think this should be written to config, in part because I worry that "default to using IPv6" will irritate some people.  But, if you want it, that's my personal preference so I'm happy to make it apply to anyone who doesn't change it.  👍

**Reply by emlowe** [reply]
We don't always save back items to config.yaml (see PR #8500). But you can do so reasonably straightforward using save_config from chia.util.config - here is an example:

```
from chia.util.config import load_config, save_config
async def update_pool_config(root_path: Path, pool_config_list: List[PoolWalletConfig]):
    full_config = load_config(root_path, "config.yaml")
    full_config["pool"]["pool_list"] = [c.to_json_dict() for c in pool_config_list]
    save_config(root_path, "config.yaml", full_config)
```

**Comment by BrandtH22** [comment]
@cross , since this ticket was opened a number of years ago without further comments it seems this issue has been resolved. 
We will be closing this ticket but if we have closed this ticket in error do not hesitate to reach out to us again with any followup questions or comments.

The best place to reach our support team is on Discord (https://discord.gg/chia) or by reopening this ticket.
